### ISSUE TITLE
load (small) bed table from memory (not http) in igv.js?

### ISSUE THREAD

**Issue by paul-shannon** [igvteam/igv.js#705]
In igvR  we often create small bed or bedGraph data.frames in R describing features of interest, then display them in igv.js by

  - writing the bed or bedGraph data.frame to a temporary file, from R, into the current working directory
  - send the temporary file path via websocket to the browser
  -  create a config record in Javascript
  - invoke igvBrowser.loadTrack(config)

Though RStudio's shiny uses the same small webserver, provided by the same httpuv R package, it adds some controls which prevent it being used to load a local file.   I have submitted a question on this to the RStudio community - maybe there is a means to circumvent those controls.

Meanwhile, may I ask if there is an igv.loadTrack config option which accepts immediate data, a properly formatted Javascript data structure with the right columns (chrom, start, end, ...)

Thanks!

 - Paul

**Comment by jrobinso** [comment]
In fact you can do that.   Its not been fully tested or documented, it was introduced for "seg" tracks but should work for any.   How many columns is your bed file?  If it does not include the exons its easy.

To directly supply features leave off the "url" property of the track description and include a "features" property  with an array of features.   For simple bed files the structure is 

```
[{chr: string,  start: number, end: number, name: string},...]
```
For bedgraph add a "value" property.  Another useful property for any type is "color".   I will be documenting these.

All other properties of a track remain the same, this just replaces "url" with an explicit array of features. This works for any ascii feature format.   It does not yet work for bam files.    See the file "examples/embedFeatures.html" for an example.

Another in memory option, more complex, is to use data uris.   These can be used for any file type.  A data uri is the complete file contents compressed and base 64 encoded.   Specifically, the file contents are zlib (gzip) compressed, then the resulting bytes base64 coded.   The resulting data URI is used in place of a URL.    For an example see examples/dataURL.html.     This was introduced to produce totally self contained html files.   Its not documented yet but is part of the 2.0 release.

**Comment by jrobinso** [comment]
Forgot to add, the internal coordinate system for all features is "bed" style,  0 based and half closed.   So the region spanning the first base of a chromosome is  0-1.

**Comment by paul-shannon** [comment]
Great! I will try it out.  Thank you   @jrobinso

 - Paul

**Comment by jrobinso** [comment]
The objects can have more properties than those specified,  everything else will go to popup text

**Comment by jrobinso** [comment]
This should be fixed

**Comment by paul-shannon** [comment]
Works great!   bed, bedGraph, seg, all three.  Thanks, Jim!

**Comment by omarwagih** [comment]
If I use the following dummy features:

```
{
    chr: "1",
    start: 3218610,
    end: 3218650,
    strand:"+",
    name: "TCGA-OR-A5J2-01"
},
{
    chr: "1",
    start: 3218630,
    end: 3218690,
    strand:"-",
    name: "TCGA-OR-A5J2-01"
}
```

I end up with the features being drawn overlapping one another:
<img width="1265" alt="Screen Shot 2019-09-08 at 3 28 20 AM" src="https://user-images.githubusercontent.com/1717767/64485030-e8cede00-d1e8-11e9-947d-a1f5bd27bd30.png">


Any idea why this is happening? Maybe I'm missing something here but I'm using the exact same code from [https://igv.org/web/release/2.2.17/examples/embedFeatures.html](https://igv.org/web/release/2.2.17/examples/embedFeatures.html), changing type to "annotation" and the feature data to what is above.

**Comment by jrobinso** [comment]
they do overlap each other, look at the coordinates

**Comment by omarwagih** [comment]
Yes, but I was expecting them to stagger vertically rather than collapse into a single row. I.e.

——————-
                ———————-

**Comment by jrobinso** [comment]
use EXPAND  displayMode.  see wiki doc for track settings.

**Comment by omarwagih** [comment]
I have with no luck on getting them to expand. This is the code I'm running:

```
 var igvDiv = document.getElementById("igv-div");
        var options = {
            genome: "hg19", 
            locus: view_interval,
            tracks:[{
                        name: "NexusHub", 
                        type: "annotation",
                        format: "bed",
                        expandedRowHeight: 100,
                        searchable: true,
                        displayMode: "EXPANDED",
                        features: window.features,
                        indexed: false
                    }]
            };
        igv.createBrowser(igvDiv, options);
```

`features` is set to the same features shared above.

**Comment by jrobinso** [comment]
You are correct, it looks like the row packing step is skipped for embedded features.  Will fix that.  Your configuration is correct.
